src/app/stores/useUserSessionStore.ts

```diff
@@ -0,0 +1,122 @@
+import {
+  browserLocalPersistence,
+  GoogleAuthProvider,
+  onAuthStateChanged,
+  setPersistence,
+  signInWithPopup,
+  signOut,
+} from "firebase/auth";
+import { auth } from "@project/firebase";
+import { addUser, getUser } from "@services/client/user";
+
+interface Session {
+  isAuthenticated: boolean;
+  isLoading: boolean;
+  error: string | null;
+}
+
+interface UserSessionState {
+  user: IdentifiableUser | null;
+  session: Session;
+  onSignIn: () => Promise<void>;
+  onSignOut: () => Promise<void>;
+  setUser: (user: IdentifiableUser | null) => void;
+  setSession: (session: Session) => void;
+}
+
+export const useUserSessionStore = create<UserSessionState>((set, get) => ({
+  user: null,
+  session: {
+    isAuthenticated: false,
+    isLoading: true,
+    error: null,
+  },
+  setUser: (user) => set({ user }),
+  setSession: (session) => set({ session }),
+  onSignIn: async () => {
+    try {
+      await setPersistence(auth, browserLocalPersistence);
+      const provider = new GoogleAuthProvider();
+      const result = await signInWithPopup(auth, provider);
+      const user = result.user;
+
+      if ((await getUser(user.uid)) == null) {
+        await addUser({
+          id: user.uid,
+          name: user.displayName ?? "",
+          tufts_username: "",
+          email: user.email ?? "",
+          role: "user",
+          courses: [],
+        });
+      }
+
+      set((state) => ({
+        session: {
+          ...state.session,
+          isLoading: true,
+        },
+      }));
+
+      window.location.href = "/private/course";
+    } catch (e: any) {
+      set({
+        session: {
+          isAuthenticated: false,
+          isLoading: false,
+          error: e.message,
+        },
+      });
+    }
+  },
+  onSignOut: async () => {
+    try {
+      set((state) => ({
+        session: { ...state.session, isLoading: true },
+      }));
+      await signOut(auth);
+      set({
+        user: null,
+        session: {
+          isAuthenticated: false,
+          isLoading: false,
+          error: null,
+        },
+      });
+
+      window.location.href = "/";
+    } catch (e: any) {
+      set({
+        session: {
+          isAuthenticated: false,
+          isLoading: false,
+          error: e.message,
+        },
+      });
+    }
+  },
+}));
+
+// Initialize auth state listener
+if (typeof window !== "undefined") {
+  onAuthStateChanged(auth, async (firebaseUser) => {
+    if (firebaseUser) {
+      const currUser = await getUser(firebaseUser?.uid);
+      useUserSessionStore.getState().setUser(currUser);
+      useUserSessionStore.getState().setSession({
+        isAuthenticated: true,
+        isLoading: false,
+        error: null,
+      });
+    } else {
+      useUserSessionStore.getState().setUser(null);
+      useUserSessionStore.getState().setSession({
+        isAuthenticated: false,
+        isLoading: false,
+        error: null,
+      });
+    }
+  });
+}
```

Comment:
I'm not sure if this auth state listener would operate on the same context as the above above store. I think it should but maybe we can also be consistent with the OH store and have this auth listener as a initlistener state? What do you think?

Reply:
It would definitely be nicer and cleaner but then we'd have to invoke it at some point before rendering. If you look at `/private/course/[courseId]/layout.tsx`:

```Javascript
React.useEffect(() => {
    const cleanup = initializeListeners(courseId);
    return () => cleanup();
  }, [courseId, initializeListeners]);
```

I reasoned that we should only call the listener when page is navigated to but with auth, I thought it'd be best if listener is invoked when the store is created.

Also, I had to make `layout.tsx` a client component but I don't know if there's any drawbacks in doing so.

 Please me know though, I could be missing something.

Reply:
Ah I see, making `/private/course/[courseId]/layout.tsx` a client component would increase page load times (because of more client sided hydration) to our core pages. 

I'm now leaning towards moving the OH initializing listeners to occur on store creation.

Reply:
Mmmm, then would we need to initializing all courses listeners on creation?